### ISSUE TITLE
How to modify YOLOv11-Pose to output keypoints with shape `(17, 4)` including visibility classification?

### ISSUE THREAD

**Issue by romanmaznikov1** [ultralytics/ultralytics#22405]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### 🧩 Question: How to modify YOLOv11-Pose to output keypoints with shape `(17, 4)` including visibility classification?

Hi Ultralytics team 👋

I’m currently experimenting with a **custom YOLOv11-Pose model**, and I’d like to modify its **keypoint head** to output an extra visibility label per keypoint.

---

#### 🎯 Goal
I want the final keypoints tensor to have the shape: `(N, 17, 4)`
instead of the standard `(N, 17, 3)`  
where each keypoint includes: `[x, y, confidence, visible_class]`

The new `visible_class` should represent:
- `0` → not visible (occluded / out of frame)  
- `1` → visible  
- `2` → uncertain / partially visible 
#### ⚙️ What I’ve tried
I’ve explored the following parts of the codebase:
- `ultralytics/models/yolo/pose/head.py` — custom `Pose` head  
- `ultralytics/models/yolo/pose/predict.py` — decoding logic (`kpts_decode`)  
- `ultralytics/nn/modules/head.py` — `Pose` class and detection head outputs  

But it’s still unclear which changes are **required and sufficient** to:
1. Extend the model head to predict 4 channels per keypoint instead of 3  
2. Correctly decode and postprocess this new fourth value  
3. Ensure training still works (loss calculation for visibility classification)

---

#### ❓ My question
Could you please clarify which components of the code need to be modified or subclassed to support **custom keypoint output shapes** (like `(N, 17, 4)`)?  
In particular:
- Where to add the extra channel in the detection head?
- Do I need to modify the loss function or data format for training?
- How should the `kpts_decode` function be adapted for the extra value?

---

#### 💡 Context
I’m training on a **custom human pose dataset** with additional visibility annotations, so having a fourth output channel for visibility would be extremely useful for downstream tasks (e.g., occlusion handling, partial body estimation).

---

Thanks a lot for your amazing work on YOLOv11-Pose 🙌  
Any guidance or example references would be greatly appreciated!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @romanmaznikov1, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (such as modifying YOLO11-Pose outputs), please provide as much information as possible so we can guide you effectively:
- Your target output schema (e.g., N x 17 x 4 with a visibility label) and how you intend to use it✅
- A brief summary of code areas you modified or plan to modify (files/functions), plus small code snippets if possible🧩
- A few sample annotations/images that show how visibility is labeled in your dataset🖼️
- Training/validation logs and any errors or mismatches you encounter📄
- Confirmation that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the detailed question—this is doable, but it touches the head, loss, and result plumbing that currently assume 17×3. Here’s the minimal, supported path:

1) Predict extra visibility as 3 logits per keypoint and keep API results 17×3
- Duplicate `yolo11n-pose.yaml` and set `kpt_shape: [17, 6]` so each keypoint is `[x, y, conf, v0, v1, v2]`. The model will honor `kpt_shape` via `PoseModel` and propagate it through the head. See the Pose model reference and how `kpt_shape` is applied in the constructor in [PoseModel](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.PoseModel).
- The pose head already computes `nk = kpt_shape[0] * kpt_shape[1]` and will emit the extra channels; only the decode needs a tiny generalization. In `ultralytics/nn/modules/head.py` `Pose.kpts_decode`, keep the current coordinate decode but apply the sigmoid only to the confidence channel (index 2), leaving your `[v0,v1,v2]` logits untouched. Reference: [Pose head and kpts_decode](https://docs.ultralytics.com/reference/nn/modules/head/#ultralytics.nn.modules.head.Pose).

Example tweak in `Pose.kpts_decode` (conf-only activation):
```python
# inside Pose.kpts_decode(...)
ndim = self.kpt_shape[1]
y = kpts.clone()
if ndim >= 3:
    if NOT_MACOS14:
        y[:, 2::ndim].sigmoid_()     # conf only
    else:
        y[:, 2::ndim] = y[:, 2::ndim].sigmoid()
y[:, 0::ndim] = (y[:, 0::ndim] * 2.0 + (self.anchors[0] - 0.5)) * self.strides
y[:, 1::ndim] = (y[:, 1::ndim] * 2.0 + (self.anchors[1] - 0.5)) * self.strides
return y
```

2) Add a CE loss for visibility classes
- Extend the pose loss to compute cross-entropy on your `[v0,v1,v2]` per keypoint while keeping the existing BCE for keypoint presence (`conf`). See [v8PoseLoss](https://docs.ultralytics.com/reference/utils/loss/#ultralytics.utils.loss.v8PoseLoss) and where keypoint losses are computed in [`calculate_keypoints_loss`](https://docs.ultralytics.com/reference/utils/loss/#ultralytics.utils.loss.v8PoseLoss.calculate_keypoints_loss).

Minimal patch inside `calculate_keypoints_loss`:
```python
# after gt_kpt, pred_kpt, kpt_mask are defined
import torch.nn.functional as F

# existing presence loss (keep)
if pred_kpt.shape[-1] >= 3:
    kpts_obj_loss = self.bce_pose(pred_kpt[..., 2], kpt_mask.float())
else:
    kpts_obj_loss = 0.0 * pred_kpt[..., 0].sum()

# new visibility CE when logits are present at [..., 3:6]
kpts_vis_loss = 0.0
if pred_kpt.shape[-1] >= 6:
    vis_logits = pred_kpt[..., 3:6].reshape(-1, 3)
    vis_targets = gt_kpt[..., 3].long().reshape(-1)  # 0/1/2 labels from your dataset
    ce_mask = kpt_mask.reshape(-1)
    if ce_mask.any():
        kpts_vis_loss = F.cross_entropy(vis_logits[ce_mask], vis_targets[ce_mask])

return kpts_loss, kpts_obj_loss + kpts_vis_loss
```

3) Keep Results keypoints at 17×3; attach visibility separately
- The `Results.Keypoints` and plotting utilities assume 17×2 or 17×3. To avoid breaking them, keep `keypoints` as `[x,y,conf]` and attach the visibility class as an extra field in your predictor override. See how predictions are assembled in [`construct_result`](https://docs.ultralytics.com/reference/models/yolo/pose/predict/#construct_result).
```python
# subclass Pose predictor
from ultralytics.models.yolo.pose.predict import PosePredictor

class PosePredictorWithVis(PosePredictor):
    def construct_result(self, pred, img, orig_img, img_path):
        result = super().construct_result(pred, img, orig_img, img_path)
        kpts = pred[:, 6:].view(pred.shape[0], *self.model.kpt_shape)  # (N, 17, 6)
        vis = kpts[..., 3:6].softmax(-1).argmax(-1)  # (N, 17) in {0,1,2}
        setattr(result, "kpt_vis", vis)              # keep result.keypoints as 17x3
        return result
```

4) Labels and validation
- Your dataset must provide the fourth value per keypoint as an integer class id `0/1/2`. The loader will feed it through as part of `keypoints`, and the loss patch above will consume it from `gt_kpt[..., 3]`.
- COCO JSON export, validators, and `plotting.kpts()` expect 17×3; if you rely on them, keep exports to 17×3 and manage visibility separately. References: [pose validator postprocess/pred_to_json](https://docs.ultralytics.com/reference/models/yolo/pose/val/#postprocess) and [plotting.kpts](https://docs.ultralytics.com/reference/utils/plotting/#ultralytics.utils.plotting.Plots.kpts).

If you get stuck, please share a minimal diff (head/loss changes only) plus a tiny sample label set, and we’ll take a look here. We can’t offer private support or partnership requests, but a focused PR with tests and docs is welcome.

**Comment by romanmaznikov1** [comment]
@glenn-jocher Thanks! It was useful for me.

**Comment by glenn-jocher** [comment]
Great to hear—if you later want to persist visibility per keypoint in text exports, override the pose val `save_one_txt` hook to append your classes; see [save_one_txt in pose val](https://docs.ultralytics.com/reference/models/yolo/pose/val/#save_one_txt). If anything breaks, please share a minimal diff + tiny sample labels, full traceback, and `yolo checks` output so we can reproduce.
